In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
energy = pd.read_csv('energy.csv')

In [3]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3649 entries, 0 to 3648
Data columns (total 21 columns):
 #   Column                                                            Non-Null Count  Dtype  
---  ------                                                            --------------  -----  
 0   Entity                                                            3649 non-null   object 
 1   Year                                                              3649 non-null   int64  
 2   Access to electricity (% of population)                           3639 non-null   float64
 3   Access to clean fuels for cooking                                 3480 non-null   float64
 4   Renewable-electricity-generating-capacity-per-capita              2718 non-null   float64
 5   Financial flows to developing countries (US $)                    1560 non-null   float64
 6   Renewable energy share in the total final energy consumption (%)  3455 non-null   float64
 7   Electricity from fossil fuels (TW

In [4]:
energy.rename(columns={'Entity': 'country',
                       'Year': 'year',
                       'Access to electricity (% of population)': 'electricity_access_%',
                       'Access to clean fuels for cooking': 'clean_fuels_access_%',
                       'Renewable-electricity-generating-capacity-per-capita':'electricity_renewable_capacity_per_capita',
                       'Financial flows to developing countries (US $)' : 'financial_flows',
                       'Renewable energy share in the total final energy consumption (%)': 'renewable_energy_share_final',
                       'Electricity from fossil fuels (TWh)': 'electricity_fossil_fuels_output',
                       'Electricity from nuclear (TWh)': 'electricity_nuclear_output',
                       'Electricity from renewables (TWh)': 'electricity_renewables_output',
                       'Low-carbon electricity (% electricity)': 'electricity_low_carbon',
                       'Primary energy consumption per capita (kWh/person)': 'energy_consumption_per_capita',
                       'Energy intensity level of primary energy (MJ/$2017 PPP GDP)': 'energy_usage_per_gdp',
                       'Value_co2_emissions_kt_by_country': 'co2_emissions_per_capita',
                       'GDP growth (annual %)': 'gdp_growth',
                       'Renewables (% equivalent primary energy)': 'renewable_energy_share_primary',
                       'Density(P/Km2)': 'density',
                       'Land Area(Km2)': 'land_area',
                       'Latitude': 'lat',
                       'Longitude': 'lon'}, inplace=True)

In [5]:
energy.shape

(3649, 21)

In [6]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)

In [7]:
missing

count          %
renewable_energy_share_primary              2137  58.563990
financial_flows                             2089  57.248561
electricity_renewable_capacity_per_capita    931  25.513839
co2_emissions_per_capita                     428  11.729241
gdp_growth                                   317   8.687312
gdp_per_capita                               282   7.728145
energy_usage_per_gdp                         207   5.672787
renewable_energy_share_final                 194   5.316525
clean_fuels_access_%                         169   4.631406
electricity_nuclear_output                   126   3.453001
electricity_low_carbon                        42   1.151000
electricity_renewables_output                 21   0.575500
electricity_fossil_fuels_output               21   0.575500
electricity_access_%                          10   0.274048
lat                                            1   0.027405
land_area                                      1   0.027405
density                                        1   0.027405
lon                                            1   0.027405
energy_consumption_per_capita                  0   0.000000
year                                           0   0.000000
country                                        0   0.000000

In [8]:
missing.sum()

count    6978.000000
%         191.230474
dtype: float64

renewable_energy_share_primary_% and financial_flows have the most missing data.

In [9]:
missing_renewables = energy[energy['renewable_energy_share_primary'].isnull()]

In [10]:
missing_renewables['country'].nunique()

104

In [11]:
missing_renewables['country'].unique()

array(['Afghanistan', 'Albania', 'Angola', 'Antigua and Barbuda',
       'Armenia', 'Aruba', 'Bahamas', 'Bahrain', 'Barbados', 'Belize',
       'Benin', 'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'Botswana',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Cayman Islands', 'Central African Republic', 'Chad', 'Comoros',
       'Congo', 'Costa Rica', 'Cuba', 'Djibouti', 'Dominica',
       'Dominican Republic', 'El Salvador', 'Equatorial Guinea',
       'Eritrea', 'Eswatini', 'Ethiopia', 'Fiji', 'French Guiana',
       'Gabon', 'Gambia', 'Georgia', 'Ghana', 'Grenada', 'Guatemala',
       'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras',
       'Jamaica', 'Jordan', 'Kenya', 'Kiribati', 'Kyrgyzstan', 'Lebanon',
       'Lesotho', 'Liberia', 'Libya', 'Madagascar', 'Malawi', 'Maldives',
       'Mali', 'Malta', 'Mauritania', 'Mauritius', 'Mongolia',
       'Montenegro', 'Mozambique', 'Myanmar', 'Namibia', 'Nauru', 'Nepal',
       'New Caledonia', 'Nicaragua', 'Niger

In [12]:
non_missing_renewables = energy[~energy['renewable_energy_share_primary'].isnull()]

In [13]:
non_missing_renewables['country'].nunique()

72

In [14]:
non_missing_renewables['country'].unique()

array(['Algeria', 'Argentina', 'Australia', 'Austria', 'Azerbaijan',
       'Bangladesh', 'Belarus', 'Belgium', 'Brazil', 'Bulgaria', 'Canada',
       'Chile', 'China', 'Colombia', 'Croatia', 'Cyprus', 'Czechia',
       'Denmark', 'Ecuador', 'Egypt', 'Estonia', 'Finland', 'France',
       'Germany', 'Greece', 'Hungary', 'Iceland', 'India', 'Indonesia',
       'Iraq', 'Ireland', 'Israel', 'Italy', 'Japan', 'Kazakhstan',
       'Kuwait', 'Latvia', 'Lithuania', 'Luxembourg', 'Malaysia',
       'Mexico', 'Morocco', 'Netherlands', 'New Zealand',
       'North Macedonia', 'Norway', 'Oman', 'Pakistan', 'Peru',
       'Philippines', 'Poland', 'Portugal', 'Qatar', 'Romania',
       'Saudi Arabia', 'Singapore', 'Slovakia', 'Slovenia',
       'South Africa', 'Spain', 'Sri Lanka', 'Sweden', 'Switzerland',
       'Thailand', 'Trinidad and Tobago', 'Turkey', 'Turkmenistan',
       'Ukraine', 'United Arab Emirates', 'United Kingdom',
       'United States', 'Uzbekistan'], dtype=object)

For renewable_energy_share_primary_, there are 104 countries with this value missing, and 72 who have this value intact. The feature is defined as "Equivalent primary energy that is derived from renewable sources." This metric considers the proportion of renewable energy sources (like wind, solar, hydro, biomass) in the total primary energy supply. It looks at how much of the raw energy extracted from natural resources comes from renewables, before processing. This feature showcases how efficienctly a country is able to use their renewable resources, which could affect the target of GDP per capita. I don't think this data is represented through any other feature in the set. Therefore, I will not include the countries who are missing this feature in the final dataset, and only focus on the 72 who have it.


In [15]:
missing_financial_flows = energy[energy['financial_flows'].isnull()]

In [16]:
non_missing_financial_flows = energy[~energy['financial_flows'].isnull()]

In [17]:
missing_financial_flows['country'].nunique()

176

In [18]:
missing_financial_flows['country'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Angola',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bermuda', 'Bhutan', 'Bosnia and Herzegovina', 'Botswana',
       'Brazil', 'Bulgaria', 'Burkina Faso', 'Burundi', 'Cambodia',
       'Cameroon', 'Canada', 'Cayman Islands', 'Central African Republic',
       'Chad', 'Chile', 'China', 'Colombia', 'Comoros', 'Congo',
       'Costa Rica', 'Croatia', 'Cuba', 'Cyprus', 'Czechia', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethiopia', 'Fiji', 'Finland', 'France',
       'French Guiana', 'Gabon', 'Gambia', 'Georgia', 'Germany', 'Ghana',
       'Greece', 'Grenada', 'Guatemala', 'Guinea', 'Guinea-Bissau',
       'Guyana', 'Haiti', 'Honduras', 'Hungary', 'Icelan

In [19]:
non_missing_financial_flows['country'].nunique()

118

In [20]:
non_missing_financial_flows['country'].unique()

array(['Afghanistan', 'Algeria', 'Angola', 'Antigua and Barbuda',
       'Argentina', 'Armenia', 'Azerbaijan', 'Bahamas', 'Bangladesh',
       'Barbados', 'Belize', 'Benin', 'Bhutan', 'Botswana', 'Brazil',
       'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon',
       'Central African Republic', 'Chad', 'Chile', 'China', 'Colombia',
       'Comoros', 'Congo', 'Costa Rica', 'Cuba', 'Djibouti', 'Dominica',
       'Dominican Republic', 'Ecuador', 'Egypt', 'El Salvador',
       'Equatorial Guinea', 'Eritrea', 'Eswatini', 'Ethiopia', 'Fiji',
       'Gabon', 'Gambia', 'Georgia', 'Ghana', 'Grenada', 'Guatemala',
       'Guinea', 'Guinea-Bissau', 'Guyana', 'Haiti', 'Honduras', 'India',
       'Indonesia', 'Iraq', 'Jamaica', 'Jordan', 'Kazakhstan', 'Kenya',
       'Kiribati', 'Kyrgyzstan', 'Lebanon', 'Lesotho', 'Liberia',
       'Madagascar', 'Malawi', 'Malaysia', 'Maldives', 'Mali',
       'Mauritania', 'Mauritius', 'Mexico', 'Mongolia', 'Morocco',
       'Mozambique', 'Myanmar', 'Namibia', 

For financial_flows, all countries have missing values, some more than others. I decide to drop this feature from the dataset as the irregularity for which it appears could skew analysis results.


In [21]:
energy.dropna(subset=["renewable_energy_share_primary"], inplace=True)

In [22]:
energy.drop(columns = 'financial_flows', inplace = True)

In [23]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1512 entries, 42 to 3564
Data columns (total 20 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   country                                    1512 non-null   object 
 1   year                                       1512 non-null   int64  
 2   electricity_access_%                       1512 non-null   float64
 3   clean_fuels_access_%                       1491 non-null   float64
 4   electricity_renewable_capacity_per_capita  693 non-null    float64
 5   renewable_energy_share_final               1440 non-null   float64
 6   electricity_fossil_fuels_output            1512 non-null   float64
 7   electricity_nuclear_output                 1407 non-null   float64
 8   electricity_renewables_output              1512 non-null   float64
 9   electricity_low_carbon                     1512 non-null   float64
 10  energy_consumption_per_capit

The final feature with a significant amount of missing values is electricity_renewable_capacity_per_capita, defined as "Installed Renewable energy capacity per person".

In [24]:
missing_renewable_capacity = energy[energy['electricity_renewable_capacity_per_capita'].isnull()]

In [25]:
non_missing_renewable_capacity = energy[~energy['electricity_renewable_capacity_per_capita'].isnull()]

In [26]:
missing_renewable_capacity['country'].nunique()

39

In [27]:
missing_renewable_capacity['country'].unique()

array(['Australia', 'Austria', 'Belarus', 'Belgium', 'Bulgaria', 'Canada',
       'Croatia', 'Cyprus', 'Czechia', 'Denmark', 'Estonia', 'Finland',
       'France', 'Germany', 'Greece', 'Hungary', 'Iceland', 'Ireland',
       'Israel', 'Italy', 'Japan', 'Latvia', 'Lithuania', 'Luxembourg',
       'Netherlands', 'New Zealand', 'North Macedonia', 'Norway',
       'Poland', 'Portugal', 'Romania', 'Slovakia', 'Slovenia', 'Spain',
       'Sweden', 'Switzerland', 'Ukraine', 'United Kingdom',
       'United States'], dtype=object)

In [28]:
non_missing_renewable_capacity['country'].nunique()

33

In [29]:
non_missing_renewable_capacity['country'].unique()

array(['Algeria', 'Argentina', 'Azerbaijan', 'Bangladesh', 'Brazil',
       'Chile', 'China', 'Colombia', 'Ecuador', 'Egypt', 'India',
       'Indonesia', 'Iraq', 'Kazakhstan', 'Kuwait', 'Malaysia', 'Mexico',
       'Morocco', 'Oman', 'Pakistan', 'Peru', 'Philippines', 'Qatar',
       'Saudi Arabia', 'Singapore', 'South Africa', 'Sri Lanka',
       'Thailand', 'Trinidad and Tobago', 'Turkey', 'Turkmenistan',
       'United Arab Emirates', 'Uzbekistan'], dtype=object)

There is a relatively even split between countries who have this feature and countries who do not, at 33 and 39 of the chosen 72. This feature is somewhat redundant, because total electricity from renewables already exists. For now, I will remove this column, and can always find population data of countries to create per capita features as I see fit.

In [30]:
energy.drop(columns = 'electricity_renewable_capacity_per_capita', inplace = True)

In [31]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1512 entries, 42 to 3564
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   country                          1512 non-null   object 
 1   year                             1512 non-null   int64  
 2   electricity_access_%             1512 non-null   float64
 3   clean_fuels_access_%             1491 non-null   float64
 4   renewable_energy_share_final     1440 non-null   float64
 5   electricity_fossil_fuels_output  1512 non-null   float64
 6   electricity_nuclear_output       1407 non-null   float64
 7   electricity_renewables_output    1512 non-null   float64
 8   electricity_low_carbon           1512 non-null   float64
 9   energy_consumption_per_capita    1512 non-null   float64
 10  energy_usage_per_gdp             1440 non-null   float64
 11  co2_emissions_per_capita         1380 non-null   float64
 12  renewable_energy_share_p

After dealing with major missing data issues, the difference between the fully intact columns vs imperfect columns is 132 rows (1512 - 1380). This difference is not significant enough to where dropping all nan rows would cripple the dataset. Still, I will keep that as a last resort and look into various forms of data imputation. Imputation can take on many forms, and will vary on a feature to feature basis.

In [32]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)

In [33]:
missing

count         %
co2_emissions_per_capita           132  8.730159
electricity_nuclear_output         105  6.944444
renewable_energy_share_final        72  4.761905
energy_usage_per_gdp                72  4.761905
gdp_growth                          65  4.298942
gdp_per_capita                      64  4.232804
clean_fuels_access_%                21  1.388889
country                              0  0.000000
renewable_energy_share_primary       0  0.000000
lat                                  0  0.000000
land_area                            0  0.000000
density                              0  0.000000
energy_consumption_per_capita        0  0.000000
year                                 0  0.000000
electricity_low_carbon               0  0.000000
electricity_renewables_output        0  0.000000
electricity_fossil_fuels_output      0  0.000000
electricity_access_%                 0  0.000000
lon                                  0  0.000000

value_co2_emissions has 132 missing values.

In [34]:
missing_counts = energy.groupby('country')['co2_emissions_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0     Slovakia                    21
1       Turkey                    21
2        Egypt                    21
3      Algeria                     1
4     Pakistan                     1
..         ...                   ...
67      Greece                     1
68     Germany                     1
69      France                     1
70     Finland                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


Egypt, Turkey, and Slovakia are all missing this feature for the entire timeframe. A quick google search links to https://data.worldbank.org/indicator/EN.ATM.CO2E.KT?end=2020&locations=SK&name_desc=false&start=1990&view=chart, which is a primary source for the dataset.

It seems the data is available for these countries, but under different names than what the author probably expected to see:

1. Egypt, Arab Republic
2. Turkiye
3. Slovak Republic

These are not the typical names associated with these countries in the US (to my experience), showing how information can be lost due to semantics. I manually add in the data for these three countries.

In [35]:
co2emissions = pd.read_csv("co2emissions.csv")
countries_to_filter = ['Turkiye', 'Egypt, Arab Rep.', 'Slovak Republic']
co2_add_ons = co2emissions[co2emissions['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = co2_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='CO2 Emission')
sorted_co2 = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_co2['Country Name'] = sorted_co2['Country Name'].replace('Turkiye', 'Turkey')
sorted_co2 = sorted_co2.rename(columns={'Country Name': 'Entity'})
sorted_co2 = sorted_co2.rename(columns={'CO2 Emission': 'Value_co2_emissions_kt_by_country'})
sorted_co2['Year'] = sorted_co2['Year'].astype(int)

In [36]:
sorted_co2

Entity  Year  Value_co2_emissions_kt_by_country
0    Egypt  2000                           114614.4
3    Egypt  2001                           126704.4
6    Egypt  2002                           129441.2
9    Egypt  2003                           133020.4
12   Egypt  2004                           144503.9
..     ...   ...                                ...
50  Turkey  2016                           376399.0
53  Turkey  2017                           418098.2
56  Turkey  2018                           414111.9
59  Turkey  2019                           398772.9
62  Turkey  2020                           407406.2

[63 rows x 3 columns]

In [37]:
sorted_co2.rename(columns={'Entity': 'country',
                       'Year': 'year',
                       'Value_co2_emissions_kt_by_country': 'co2_emissions_per_capita'
                      }, inplace=True)

In [38]:
energy.set_index(['country', 'year'], inplace=True)
sorted_co2.set_index(['country', 'year'], inplace=True)
energy.update(sorted_co2)
energy.reset_index(inplace=True)

In [39]:
missing_counts = energy.groupby('country')['co2_emissions_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

            country  Missing_Values_Count
0           Algeria                     1
1   North Macedonia                     1
2            Poland                     1
3       Philippines                     1
4              Peru                     1
..              ...                   ...
67          Estonia                     1
68       Uzbekistan                     1
69            Egypt                     0
70           Turkey                     0
71         Slovakia                     0

[72 rows x 2 columns]


The 63 missing rows of co2 emissions for the three countries have been taken care of. For the rest of missing emissions, I will impute with the same value from the previous year since countries arte only missing a single year.

In [40]:
energy['co2_emissions_per_capita'] = energy.groupby('country')['co2_emissions_per_capita'].fillna(method='ffill')

In [41]:
missing_counts = energy.groupby('country')['co2_emissions_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0      Algeria                     0
1    Argentina                     0
2        Qatar                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


All taken care of.

Next is electricity from nuclear, with 105 missing values.

In [42]:
missing_counts = energy.groupby('country')['electricity_nuclear_output'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0       Malaysia                    21
1   Saudi Arabia                    21
2          Chile                    21
3      Indonesia                    21
4     Kazakhstan                    21
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70        Kuwait                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


5 countries are missing values for this feature in its entirety.

In [43]:
entities_with_zero_nuclear = energy[energy['electricity_nuclear_output'] == 0]['country']
print("countries with documented 0s for nuclear output")
print(entities_with_zero_nuclear.unique())

countries with documented 0s for nuclear output
['Algeria' 'Australia' 'Austria' 'Azerbaijan' 'Bangladesh' 'Belarus'
 'Colombia' 'Croatia' 'Cyprus' 'Denmark' 'Ecuador' 'Egypt' 'Estonia'
 'Greece' 'Iceland' 'Iraq' 'Ireland' 'Israel' 'Italy' 'Japan' 'Kuwait'
 'Latvia' 'Lithuania' 'Luxembourg' 'Morocco' 'New Zealand'
 'North Macedonia' 'Norway' 'Oman' 'Peru' 'Philippines' 'Poland'
 'Portugal' 'Qatar' 'Singapore' 'Sri Lanka' 'Thailand'
 'Trinidad and Tobago' 'Turkey' 'Turkmenistan' 'United Arab Emirates'
 'Uzbekistan']


It seems these 5 countries (Malaysia, Saudi Arabia, Chile, Indonesia, Kazakhstan) are true nan values, because there are several other countries with values of 0 who are documented properly, and I don't see their names here.

I looked into the nuclear energy programs for the 5, and found the following:

1. Malaysia - "In the past, the country had an ambitious nuclear program with plans to establish two operational nuclear power plants by 2021. However, these plans took an unexpected turn and were indefinitely postponed, ultimately leading to the disbandment of the Malaysia Nuclear Power Cooperation in 2018." (https://www.nuclearbusiness-platform.com/media/insights/nuclear-potentially-can-help-malaysia-to-become-clean-energy-hub#:~:text=In%20the%20past%2C%20the%20country,Nuclear%20Power%20Cooperation%20in%202018.)

2. Saudi Arabia - "Saudi Arabia has plans to establish a civil nuclear power industry. The country in 2013 projected 17 GWe of nuclear capacity by 2032, but that target was abandoned two years later. At present, virtually all of the country's electricity is generated by burning fossil fuels." (https://world-nuclear.org/information-library/country-profiles/countries-o-s/saudi-arabia)

3. Chile - "Currently, Chile has no nuclear power plants and has not taken any decision regarding the development of a nuclear power programme. However, Chile has a strong background on other nuclear applications." (https://www-pub.iaea.org/MTCD/Publications/PDF/cnpp2018/countryprofiles/Chile/Chile.htm#:~:text=Currently%2C%20Chile%20has%20no%20nuclear,background%20on%20other%20nuclear%20applications.)

4. Indonesia - as of March 2023, "The USA and Indonesia have announced a strategic partnership to help Indonesia develop its nuclear energy programme, supporting Indonesia's interest in deploying small modular reactor (SMR) technology to meet its energy security and climate goals." (https://world-nuclear-news.org/Articles/USA,-Indonesia-announce-partnership-on-SMRs)

5. Kazakhstan - as of September 2023, "Kazakhstan is to hold a referendum on the construction of a nuclear power plant in the country, President Kassym-Jomart Tokayev announced during a state-of-the-nation address. The date of the vote has yet to be determined." (https://world-nuclear-news.org/Articles/Kazakh-people-to-decide-on-nuclear-plant-construct)

These countries do not currently have operable nuclear power in their countries, but some are looking to investigate the possibility.  

Imputing the missing data with 0 makes the most sense for these 5 countries.




In [44]:
energy['electricity_nuclear_output'].fillna(0, inplace=True)

In [45]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
renewable_energy_share_final        72  4.761905
energy_usage_per_gdp                72  4.761905
gdp_growth                          65  4.298942
gdp_per_capita                      64  4.232804
clean_fuels_access_%                21  1.388889
country                              0  0.000000
co2_emissions_per_capita             0  0.000000
lat                                  0  0.000000
land_area                            0  0.000000
density                              0  0.000000
renewable_energy_share_primary       0  0.000000
energy_consumption_per_capita        0  0.000000
year                                 0  0.000000
electricity_low_carbon               0  0.000000
electricity_renewables_output        0  0.000000
electricity_nuclear_output           0  0.000000
electricity_fossil_fuels_output      0  0.000000
electricity_access_%                 0  0.000000
lon                                  0  0.000000

In [46]:
missing_counts = energy.groupby('country')['renewable_energy_share_final'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0      Algeria                     1
1    Argentina                     1
2        Qatar                     1
3     Portugal                     1
4       Poland                     1
..         ...                   ...
67     Germany                     1
68      France                     1
69     Finland                     1
70     Estonia                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


In [47]:
missing_counts = energy.groupby('country')['energy_usage_per_gdp'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0      Algeria                     1
1    Argentina                     1
2        Qatar                     1
3     Portugal                     1
4       Poland                     1
..         ...                   ...
67     Germany                     1
68      France                     1
69     Finland                     1
70     Estonia                     1
71  Uzbekistan                     1

[72 rows x 2 columns]


For these two features, it looks like just a single year is missing for every country, so I will impute with the previous value.

In [48]:
energy['renewable_energy_share_final'] = energy.groupby('country')['renewable_energy_share_final'].fillna(method='ffill')
energy['energy_usage_per_gdp'] = energy.groupby('country')['energy_usage_per_gdp'].fillna(method='ffill')

In [49]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
gdp_growth                          65  4.298942
gdp_per_capita                      64  4.232804
clean_fuels_access_%                21  1.388889
country                              0  0.000000
energy_usage_per_gdp                 0  0.000000
lat                                  0  0.000000
land_area                            0  0.000000
density                              0  0.000000
renewable_energy_share_primary       0  0.000000
co2_emissions_per_capita             0  0.000000
energy_consumption_per_capita        0  0.000000
year                                 0  0.000000
electricity_low_carbon               0  0.000000
electricity_renewables_output        0  0.000000
electricity_nuclear_output           0  0.000000
electricity_fossil_fuels_output      0  0.000000
renewable_energy_share_final         0  0.000000
electricity_access_%                 0  0.000000
lon                                  0  0.000000

Just three more imperfect features to go.

In [50]:
missing_counts = energy.groupby('country')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0          Egypt                    21
1       Slovakia                    21
2        Czechia                    21
3          Qatar                     1
4   Turkmenistan                     1
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70    Kazakhstan                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


In [51]:
missing_counts = energy.groupby('country')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0          Egypt                    21
1       Slovakia                    21
2        Czechia                    21
3   Turkmenistan                     1
4       Pakistan                     0
..           ...                   ...
67        Israel                     0
68         Italy                     0
69         Japan                     0
70    Kazakhstan                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


Once again, Egypt and Slovakia are missing data, and now Czehcia as well. I will find this information from the World Bank and update the main dataframe.

In [52]:
gdpgrowth = pd.read_csv("gdpgrowth.csv")
countries_to_filter = ['Czechia', 'Egypt, Arab Rep.', 'Slovak Republic']
gdpgrowth_add_ons = gdpgrowth[gdpgrowth['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = gdpgrowth_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='gdp_growth')
sorted_gdpgrowth = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_gdpgrowth['Country Name'] = sorted_gdpgrowth['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_gdpgrowth['Country Name'] = sorted_gdpgrowth['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_gdpgrowth = sorted_gdpgrowth.rename(columns={'Country Name': 'Entity'})
sorted_gdpgrowth['Year'] = sorted_gdpgrowth['Year'].astype(int)

In [53]:
sorted_gdpgrowth.rename(columns={'Entity': 'country',
                       'Year': 'year',
                      }, inplace=True)

In [54]:
energy.set_index(['country', 'year'], inplace=True)
sorted_gdpgrowth.set_index(['country', 'year'], inplace=True)
energy.update(sorted_gdpgrowth)
energy.reset_index(inplace=True)

In [55]:
gdpcapita = pd.read_csv("gdpcapita.csv")
countries_to_filter = ['Czechia', 'Egypt, Arab Rep.', 'Slovak Republic']
gdpcapita_add_ons = gdpcapita[gdpgrowth['Country Name'].isin(countries_to_filter)]
columns_to_keep = ['Country Name'] + [str(year) for year in range(2000, 2021)]
filtered_df = gdpcapita_add_ons.filter(columns_to_keep)
melted_df = filtered_df.melt(id_vars='Country Name', var_name='Year', value_name='gdp_per_capita')
sorted_gdpcapita = melted_df.sort_values(by=['Country Name', 'Year'])
sorted_gdpcapita['Country Name'] = sorted_gdpcapita['Country Name'].replace('Egypt, Arab Rep.', 'Egypt')
sorted_gdpcapita['Country Name'] = sorted_gdpcapita['Country Name'].replace('Slovak Republic', 'Slovakia')
sorted_gdpcapita = sorted_gdpcapita.rename(columns={'Country Name': 'Entity'})
sorted_gdpcapita['Year'] = sorted_gdpcapita['Year'].astype(int)

In [56]:
sorted_gdpcapita.rename(columns={'Entity': 'country',
                       'Year': 'year',
                      }, inplace=True)

In [57]:
energy.set_index(['country', 'year'], inplace=True)
sorted_gdpcapita.set_index(['country', 'year'], inplace=True)
energy.update(sorted_gdpcapita)
energy.reset_index(inplace=True)

In [58]:
missing_counts = energy.groupby('country')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0          Qatar                     1
1   Turkmenistan                     1
2         Norway                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67       Germany                     0
68        France                     0
69       Finland                     0
70       Estonia                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


In [59]:
missing_counts = energy.groupby('country')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0   Turkmenistan                     1
1        Algeria                     0
2         Norway                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67       Germany                     0
68        France                     0
69       Finland                     0
70       Estonia                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


Much better, I'll impute with previous values.

In [60]:
energy['gdp_growth'] = energy.groupby('country')['gdp_growth'].fillna(method='ffill')
energy['gdp_per_capita'] = energy.groupby('country')['gdp_per_capita'].fillna(method='ffill')

In [61]:
missing_counts = energy.groupby('country')['gdp_growth'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0        Qatar                     1
1      Algeria                     0
2    Argentina                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


In [62]:
missing_counts = energy.groupby('country')['gdp_per_capita'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

       country  Missing_Values_Count
0      Algeria                     0
1    Argentina                     0
2        Qatar                     0
3     Portugal                     0
4       Poland                     0
..         ...                   ...
67     Germany                     0
68      France                     0
69     Finland                     0
70     Estonia                     0
71  Uzbekistan                     0

[72 rows x 2 columns]


Strange, Qatar is still missing a single value for gdp_growth.

In [63]:
qatar = energy['country'] == 'Qatar'

In [64]:
energy[qatar]['gdp_growth']

1092          NaN
1093     3.898187
1094     7.182152
1095     3.719959
1096    19.218915
1097     7.492758
1098    26.170246
1099    17.985657
1100    17.663556
1101    11.956561
1102    19.592332
1103    13.375176
1104     4.730012
1105     5.556041
1106     5.334323
1107     4.753346
1108     3.064192
1109    -1.497605
1110     1.234872
1111     0.688241
1112    -3.557604
Name: gdp_growth, dtype: float64

That's why, I can't impute from previous because there is none, so I'll impute with a forward fill.

In [65]:
energy['gdp_growth'] = energy.groupby('country')['gdp_growth'].fillna(method='bfill')

In [66]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count         %
clean_fuels_access_%                21  1.388889
country                              0  0.000000
energy_usage_per_gdp                 0  0.000000
lat                                  0  0.000000
land_area                            0  0.000000
density                              0  0.000000
gdp_per_capita                       0  0.000000
gdp_growth                           0  0.000000
renewable_energy_share_primary       0  0.000000
co2_emissions_per_capita             0  0.000000
energy_consumption_per_capita        0  0.000000
year                                 0  0.000000
electricity_low_carbon               0  0.000000
electricity_renewables_output        0  0.000000
electricity_nuclear_output           0  0.000000
electricity_fossil_fuels_output      0  0.000000
renewable_energy_share_final         0  0.000000
electricity_access_%                 0  0.000000
lon                                  0  0.000000

Last set of missing values - 21 for clean_fuels_access_%.

In [67]:
missing_counts = energy.groupby('country')['clean_fuels_access_%'].apply(lambda x: x.isnull().sum()).sort_values(ascending = False)
sorted_entities = missing_counts.reset_index(name='Missing_Values_Count')
print(sorted_entities)

         country  Missing_Values_Count
0       Bulgaria                    21
1   Saudi Arabia                     0
2          Qatar                     0
3       Portugal                     0
4         Poland                     0
..           ...                   ...
67        Greece                     0
68       Germany                     0
69        France                     0
70       Finland                     0
71    Uzbekistan                     0

[72 rows x 2 columns]


All from Bulgaria, world bank is missing the metric for the country in its entirety, so I'll drop Bulgaria from the dataset.

In [68]:
energy = energy[energy['country'] != 'Bulgaria']

In [69]:
missing = pd.concat([energy.isnull().sum(), 100 * energy.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing = missing.sort_values(by='%', ascending=False)
missing

count    %
country                              0  0.0
energy_usage_per_gdp                 0  0.0
lat                                  0  0.0
land_area                            0  0.0
density                              0  0.0
gdp_per_capita                       0  0.0
gdp_growth                           0  0.0
renewable_energy_share_primary       0  0.0
co2_emissions_per_capita             0  0.0
energy_consumption_per_capita        0  0.0
year                                 0  0.0
electricity_low_carbon               0  0.0
electricity_renewables_output        0  0.0
electricity_nuclear_output           0  0.0
electricity_fossil_fuels_output      0  0.0
renewable_energy_share_final         0  0.0
clean_fuels_access_%                 0  0.0
electricity_access_%                 0  0.0
lon                                  0  0.0

No more missing values.

In [70]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1491 entries, 0 to 1511
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   country                          1491 non-null   object 
 1   year                             1491 non-null   int64  
 2   electricity_access_%             1491 non-null   float64
 3   clean_fuels_access_%             1491 non-null   float64
 4   renewable_energy_share_final     1491 non-null   float64
 5   electricity_fossil_fuels_output  1491 non-null   float64
 6   electricity_nuclear_output       1491 non-null   float64
 7   electricity_renewables_output    1491 non-null   float64
 8   electricity_low_carbon           1491 non-null   float64
 9   energy_consumption_per_capita    1491 non-null   float64
 10  energy_usage_per_gdp             1491 non-null   float64
 11  co2_emissions_per_capita         1491 non-null   float64
 12  renewable_energy_share_pr

In [71]:
energy['density'] = energy['density'].str.replace(',', '').astype(int)

In [72]:
energy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1491 entries, 0 to 1511
Data columns (total 19 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   country                          1491 non-null   object 
 1   year                             1491 non-null   int64  
 2   electricity_access_%             1491 non-null   float64
 3   clean_fuels_access_%             1491 non-null   float64
 4   renewable_energy_share_final     1491 non-null   float64
 5   electricity_fossil_fuels_output  1491 non-null   float64
 6   electricity_nuclear_output       1491 non-null   float64
 7   electricity_renewables_output    1491 non-null   float64
 8   electricity_low_carbon           1491 non-null   float64
 9   energy_consumption_per_capita    1491 non-null   float64
 10  energy_usage_per_gdp             1491 non-null   float64
 11  co2_emissions_per_capita         1491 non-null   float64
 12  renewable_energy_share_pr

In [73]:
energy.shape

(1491, 19)

Add in 3 descriptive features
land_area_category: small, medium, large <br>
density_category: sparse, populated, crowded <br>
quadrant: NW, NE, SW, SE <br>

In [74]:
quantiles_land_area = energy['land_area'].quantile([0, 1/3, 2/3, 1]).to_list()
labels_land_area = ['small', 'medium', 'large']
energy['land_area_category'] = pd.cut(energy['land_area'], bins=quantiles_land_area, labels=labels_land_area, include_lowest=True)

In [75]:
quantiles_density = energy['density'].quantile([0, 1/3, 2/3, 1]).to_list()
labels_density = ['sparse', 'populated', 'packed']
energy['density_category'] = pd.cut(energy['density'], bins=quantiles_density, labels=labels_density, include_lowest=True)

In [76]:
energy['density_category'].value_counts()

density_category
sparse       504
populated    504
packed       483
Name: count, dtype: int64

In [77]:
energy['land_area_category'].value_counts()

land_area_category
small     504
medium    504
large     483
Name: count, dtype: int64

In [78]:
def classify_quadrant(lat, lon):
    if lat > 0 and lon > 0:
        return 'NE'
    elif lat > 0 and lon < 0:
        return 'NW'
    elif lat < 0 and lon > 0:
        return 'SE'
    elif lat < 0 and lon < 0:
        return 'SW'
    else:
        return 'Unknown'

energy['quadrant'] = energy.apply(lambda row: classify_quadrant(row['lat'], row['lon']), axis=1)

In [79]:
energy.shape

(1491, 22)

In [80]:
energy

country  year  electricity_access_%  clean_fuels_access_%  \
0        Algeria  2000              98.97310                 97.10   
1        Algeria  2001              98.96687                 97.30   
2        Algeria  2002              98.95306                 97.80   
3        Algeria  2003              98.93401                 98.00   
4        Algeria  2004              98.91208                 98.20   
...          ...   ...                   ...                   ...   
1507  Uzbekistan  2016             100.00000                 85.20   
1508  Uzbekistan  2017             100.00000                 84.90   
1509  Uzbekistan  2018             100.00000                 84.30   
1510  Uzbekistan  2019             100.00000                 84.60   
1511  Uzbekistan  2020             100.00000                 84.25   

      renewable_energy_share_final  electricity_fossil_fuels_output  \
0                             0.43                            23.84   
1                             0.43                            24.96   
2                             0.51                            25.94   
3                             0.47                            27.54   
4                             0.44                            29.14   
...                            ...                              ...   
1507                          1.61                            48.75   
1508                          1.75                            49.71   
1509                          1.49                            53.58   
1510                          1.57                            53.64   
1511                          1.57                            50.93   

      electricity_nuclear_output  electricity_renewables_output  \
0                            0.0                           0.05   
1                            0.0                           0.07   
2                            0.0                           0.06   
3                            0.0                           0.26   
4                            0.0                           0.25   
...                          ...                            ...   
1507                         0.0                           7.25   
1508                         0.0                           8.35   
1509                         0.0                           5.85   
1510                         0.0                           6.47   
1511                         0.0                           5.01   

      electricity_low_carbon  energy_consumption_per_capita  ...  \
0                   0.209293                       9746.524  ...   
1                   0.279664                       9961.640  ...   
2                   0.230769                      10180.350  ...   
3                   0.935252                      10510.461  ...   
4                   0.850630                      10759.022  ...   
...                      ...                            ...  ...   
1507               12.946428                      16374.342  ...   
1508               14.381675                      16642.676  ...   
1509                9.843514                      16445.740  ...   
1510               10.763600                      16212.221  ...   
1511                8.956025                      15860.609  ...   

      renewable_energy_share_primary  gdp_growth  gdp_per_capita  density  \
0                           0.053235    3.800000     1765.027146       18   
1                           0.065218    3.000000     1740.606654       18   
2                           0.051677    5.600000     1781.828908       18   
3                           0.228104    7.200000     2103.381291       18   
4                           0.206787    4.300000     2610.185422       18   
...                              ...         ...             ...      ...   
1507                        3.567936    5.932151     2704.677188       79   
1508                        3.972285    4.395275     1916.764642       79   


In [81]:
energy.to_csv('energy_cleaned.csv', index=False)